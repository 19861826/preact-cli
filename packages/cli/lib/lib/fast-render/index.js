const { info, warn } = require('../../util');
const { join, resolve } = require('path');
const { Worker } = require('worker_threads');
const { writeFile, mkdir } = require('../../fs');
const { PRERENDER_DATA_FILE_NAME } = require('../constants');
const pool = require('./pool');

module.exports = async function fastPrerender(
	{ src, dest, cwd, prerenderUrls },
	stats
) {
	let pages = [{ url: '/' }];
	if (prerenderUrls) {
		try {
			let result = require(resolve(cwd, prerenderUrls));
			if (typeof result.default !== 'undefined') {
				result = result.default();
			}
			if (typeof result === 'function') {
				result = await result();
			}
			if (typeof result === 'string') {
				result = JSON.parse(result);
			}
			if (result instanceof Array) {
				pages = result;
			}
		} catch (error) {
			warn('Failed to load prerenderUrls file, using default!\n');
		}
	}

	// eslint-disable-next-line no-console
	console.log('\n\n');
	info(
		`Prerendering ${pages.length} page${
			pages.length > 0 && 's'
		} from ${prerenderUrls}.`
	);
	const pageData = {};
	const renderedContents = await pool(
		pages.map(data => {
			pageData[data.url] = data;
			return {
				src,
				dest,
				cwd,
				webpack: {
					publicPath: stats.compilation.outputOptions.publicPath,
					assets: Object.keys(stats.compilation.assets),
				},
				data,
			};
		}),
		({ src, dest, cwd, webpack, data }) => {
			return new Promise(resolve => {
				const worker = new Worker(join(__dirname, 'renderer.js'), {
					workerData: {
						src,
						dest,
						cwd,
						webpack,
						data,
					},
				});
				worker.once('message', async preRenderContent => {
					worker.terminate();
					resolve(preRenderContent);
				});
			});
		}
	);
	renderedContents.forEach(async preRenderedContent => {
		const dirPath = preRenderedContent.url.endsWith('.html')
			? preRenderedContent.url.substring(
					0,
					preRenderedContent.url.lastIndexOf('/')
			  )
			: preRenderedContent.url;
		const filePath = preRenderedContent.url.endsWith('.html')
			? preRenderedContent.url
			: join(preRenderedContent.url, 'index.html');
		await mkdir(join(dest, dirPath), {
			recursive: true,
		});
		await writeFile(join(dest, filePath), preRenderedContent.content);
		await writeFile(
			join(dest, dirPath, PRERENDER_DATA_FILE_NAME),
			JSON.stringify(pageData[preRenderedContent.url])
		);
	});
};
