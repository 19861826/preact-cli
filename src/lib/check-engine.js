import { execSync } from "child_process";

import semver from "semver";
import colors from "chalk";
import pkg from "./../../package.json";

const platformSuffix = process.platform === "win32" ? ".cmd" : "";
const node = "node" + platformSuffix;

export default function checkVersion() {
  const requirements = pkg.engines;
  const nodeVersion = execSync(node + " --version").toString() || '';

  if ( !(semver.satisfies(nodeVersion, requirements.node)) ) {
    const errorMessage = colors.yellow(
      "\n⚠️ " +
        "preact-cli requires at least " +
        "node@" +
        requirements.node +
        "\n\n" +
        "Your node version is " +
        nodeVersion +
      "\n"
    );
		process.stdout.write(errorMessage) + '\n';
		process.exit(1);
  }

	return true;
}
