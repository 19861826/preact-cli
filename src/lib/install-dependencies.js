import spawn from 'cross-spawn-promise';
import { commandExists } from './shell';

const initialize = async (yarn, cwd) => {
	let isYarnAvailable = await commandExists('yarn');

	if (isYarnAvailable && yarn) {
		return await spawn('yarn', ['init', '-y'], { cwd, stdio: 'ignore' });
	}

	await spawn('npm', ['init', '-y'], { cwd, stdio: 'ignore' });
};

const install = async (yarn, cwd, packages, env) => {
	let isDev = env === 'dev' ? true : false;
	let isYarnAvailable = await commandExists('yarn');
	let toInstall = packages.filter(Boolean);

	if (isYarnAvailable && yarn) {
		let args = ['add'];
		if (isDev) {
			args.push('-D');
		}

		return await spawn('yarn', [...args, ...toInstall], { cwd, stdio: 'ignore' });
	}

	await spawn('npm', ['install', isDev ? '--save-dev' : '--save', ...toInstall], { cwd, stdio: 'ignore' });
};

const pkgScripts = async (yarn, pkg) => {
	let isYarnAvailable = await commandExists('yarn');

	if (isYarnAvailable && yarn) {
		return {
			...(pkg.scripts || {}),
			start: 'if-env NODE_ENV=production && yarnpkg -s serve || yarnpkg -s dev',
			build: 'preact build',
			serve: 'preact build && preact serve',
			dev: 'preact watch',
			test: 'eslint src && preact test'
		};
	}

	return {
		...(pkg.scripts || {}),
		start: 'if-env NODE_ENV=production && npm run -s serve || npm run -s dev',
		build: 'preact build',
		serve: 'preact build && preact serve',
		dev: 'preact watch',
		test: 'eslint src && preact test'
	};
};

export { install, initialize, pkgScripts };
