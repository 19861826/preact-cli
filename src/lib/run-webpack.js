import path from 'path';
import rimraf from 'rimraf';
import fs from 'fs.promised';
import webpack from 'webpack';
import WebpackDevServer from 'webpack-dev-server';
import chalk from 'chalk';

export default (watch=false, env, config, onprogress) => {
	if (watch) {
		return devBuild(config, onprogress);
	}
	else {
		return prodBuild(env, config);
	}
};

const devBuild = (config, onprogress) => {
	let compiler = webpack(config);
	return new Promise((resolve, reject) => {
		let first = true;
		compiler.plugin('done', stats => {
			if (first) {
				first = false;
				let devServer = config.devServer;
				let serverAddr = `${devServer.https===true?'https':'http'}://${process.env.HOST || devServer.host || 'localhost'}:${process.env.PORT || devServer.port || 8080}`;
				process.stdout.write(`  \u001b[32m> Development server started at ${serverAddr}\u001b[39m\n`);
			}
			if (onprogress) onprogress(stats);
		});
		compiler.plugin('failed', reject);

		let server = new WebpackDevServer(compiler, config.devServer);
		server.listen(config.devServer.port);
	});
};

const prodBuild = (env, config) => {
	let ssrOutputPath = path.resolve(__dirname, 'ssr-build/');
	let ssrConfig = (config) => {
		let newConfig = Object.assign({}, config, {
			target: 'node',
			entry: path.resolve(env.cwd, env.src || 'src', 'index.js'),
			plugins: config.plugins.filter(c => !(c.constructor && c.constructor.name === 'HtmlWebpackPlugin')),
		});

		newConfig.output.path = ssrOutputPath;
		newConfig.output.libraryTarget = 'commonjs2';

		let asyncLoaderIndex = newConfig.module.loaders
			.findIndex(l => l.loader === path.resolve(__dirname, 'async-component-loader'));

		// Remove async loader in prerender because all chunks are available.
		if (asyncLoaderIndex > -1) {
			newConfig.module.loaders.splice(asyncLoaderIndex, 1);
		}

		return newConfig;
	};
	let compiler = webpack(config);
	let runCompiler = () => new Promise((resolve, reject) => {
		console.log('Running production build...');
		compiler.run((err, stats) => {
			if (err) reject(err);
			else {
				// Timeout for plugins that work on `after-emit` event of webpack
				setTimeout(()=>	resolve(stats), 20);
			}
		});
	});

	let ssrCompiler = webpack(ssrConfig(config));
	let runSsrCompiler = () => new Promise((resolve, reject) => {
		console.log('Running SSR build...');
		rimraf.sync(ssrOutputPath);
		ssrCompiler.run((err) => {
			if (err) return reject(err);
			// Timeout for plugins that work on `after-emit` event of webpack
			setTimeout(()=>	resolve(), 20);
		});
	});

	return env.prerender ? runSsrCompiler().then(runCompiler) : runCompiler();
};

export function showStats(stats) {
	let info = stats.toJson();

	if (stats.hasErrors()) {
		info.errors.forEach( message => {
			process.stderr.write(message+'\n');
		});
	}

	if (stats.hasWarnings()) {
		info.warnings.forEach( message => {
			process.stderr.write(chalk.yellow(message)+'\n');
		});
	}

	return stats;
}

export function writeJsonStats(stats) {
	const outputPath = path.resolve(process.cwd(), 'stats.json');
	const jsonStats = stats.toJson({
		json: true,
		chunkModules: true,
		source: false,
	});

	jsonStats.modules.forEach(normalizeModule);
	jsonStats.chunks.forEach(c => c.modules.forEach(normalizeModule));

	return fs.writeFile(outputPath, JSON.stringify(jsonStats))
		.then(() => {
			process.stdout.write('\nWebpack output stats generated.\n\n');
			process.stdout.write('You can upload your stats.json to:\n');
			process.stdout.write('- https://chrisbateman.github.io/webpack-visualizer/\n');
			process.stdout.write('- https://webpack.github.io/analyse/\n');
		});
}

const normalizeModule = m => {
	const keysToNormalize = ['identifier', 'name', 'module', 'moduleName', 'moduleIdentifier'];

	keysToNormalize.forEach(key => {
		if (key in m) {
			m[key] = normalizeName(m[key]);
		}
	});

	if (m.reasons) {
		m.reasons.forEach(normalizeModule);
	}

	return m;
};

const normalizeName = p => p.substr(p.lastIndexOf('!') + 1);
