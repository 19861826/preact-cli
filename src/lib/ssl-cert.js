const fs = require('fs.promised');
const { resolve } = require('path');
const { execFile } = require('child_process');
const persistencePath = require('persist-path');
const simplehttp2server = require('simplehttp2server');
const getDevelopmentCertificate = require('devcert-san');

function spawnServerForCert() {
	return new Promise((res, rej) => {
		let cwd = persistencePath('preact-cli');
		let child = execFile(simplehttp2server, ['-listen', ':40210'], {
			cwd,
			encoding: 'utf8'
		}, (err, stdout, stderr) => {
			if (err) return rej(err);
			let timer = setTimeout( () => {
				rej('Error: certificate generation timed out.');
			}, 5000);
			async function check(chunk) {
				if (/listening/gi.match(chunk)) {
					clearTimeout(timer);
					child.kill();
					res({
						key: await fs.readFile(resolve(cwd, 'key.pem'), 'utf-8'),
						cert: await fs.readFile(resolve(cwd, 'cert.pem'), 'utf-8'),
						keyPath: resolve(cwd, 'key.pem'),
						certPath: resolve(cwd, 'cert.pem')
					});
				}
			}
			stdout.on('data', check);
			stderr.on('data', check);
		});
	});
}

module.exports = async function () {
	process.stdout.write('Setting up SSL certificate (may require sudo)...\n');
	try {
		return await getDevelopmentCertificate('preact-cli', {
			installCertutil: true
		});
	} catch (err) {
		process.stderr.write('Attempting to spawn simplehttp2server to generate cert.\n');
		try {
			return await spawnServerForCert();
		} catch (err) {
			process.stderr.write(`Failed to generate dev SSL certificate: ${err}\n`);
		}
	}
	return false;
}
