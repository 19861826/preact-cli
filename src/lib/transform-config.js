import path from 'path'
import fs from 'fs.promised'
import {
	webpack,
} from '@webpack-blocks/webpack2';

const getRules = config => [...(config.module.loaders || []), ...(config.module.rules || [])]
	.map((rule, index) => ({ index, rule }))

const getLoaders = config => getRules(config)
	.map(({ rule, index }) => ({
		rule: rule,
		ruleIndex: index,
		loaders: (rule.loaders || rule.use || rule.loader)
	}))

const getPlugins = config => (config.plugins || []).map((plugin, index) => ({ index, plugin }))

const helpers = cwd => ({
	webpack,

	getRulesByMatchingFile(config, file) {
		let filePath = path.resolve(cwd, file)
		return getRules(config)
			.filter(w => w.rule.test && w.rule.test.exec(filePath))
	},

	getLoadersByName(config, name) {
		return getLoaders(config)
			.map(({ rule, ruleIndex, loaders }) => Array.isArray(loaders)
				? loaders.map((loader, loaderIndex) => ({ rule, ruleIndex, loader, loaderIndex }))
				: [{ rule, ruleIndex, loader: loaders, loaderIndex: -1 }]
			)
			.reduce((arr, loaders) => arr.concat(loaders), [])
			.filter(({ loader }) => loader === name || (loader && loader.loader === name))
	},

	getPluginsByName(config, name) {
		return getPlugins(config).filter(w => w.plugin && w.plugin.constructor && w.plugin.constructor.name === name)
	},

	getPluginsByType(config, type) {
		return getPlugins(config).filter(w => w.plugin instanceof type)
	},
})

export default async function (env, config) {
	const transformerPath = path.resolve(env.cwd, env.config);
	try {
		await fs.stat(transformerPath);
	} catch (e) {
		return;
	}

	require('babel-register')();
	const m = require(transformerPath);
	const transformer = m && m.default || m;
	transformer(config, Object.assign({}, env), helpers(env.cwd))
}
