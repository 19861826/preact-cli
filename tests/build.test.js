import test from './async-test';
import { resolve } from 'path';
import { create, build } from './lib/cli';
import lsr from './lib/lsr';
import { setup, clean } from './lib/output';

test('preact build - before', async () => {
	await setup();
});

['empty', 'simple', 'root', 'default'].forEach(template =>
	test(`preact build - should produce output. Veryfing ${template}`, async t => {
		let app = await create('app', template);
		await build(app);

		let output = await lsr(resolve(app, 'build'));

		t.isEquivalent(prepare(output), expectedOutputs[template]);
	})
);

test('preact build - after', async () => {
	await clean();
});

const prepare = obj => {
	let keys = Object.keys(obj);

	if (keys.length === 1 && keys[0] === 'size' && typeof keys[0] === 'number') {
		return { size: Math.round(obj.size / 10) * 10 };
	}

	return keys.reduce((agg, key) => {
		let newKey = key.replace(/\.chunk\.\w+\./, '.chunk.*.');
		agg[newKey] = prepare(obj[key]);
		return agg;
	}, {});
};


const smallBuild = {
	assets: {
		'favicon.ico': { size: 15086 },
		'icon.png': { size: 51484 }
	},
	'polyfills.chunk.*.js': { size: 4068 },
	'favicon.ico': { size: 15086 },
	'sw.js': { size: 3378 },
	'manifest.json': { size: 298 },
	'push-manifest.json': { size: 2 },
};

const fullBuild = {
	assets: {
		'favicon.ico': { size: 15086 },
		icons: {
			'android-chrome-192x192.png': { size: 14058 },
			'android-chrome-512x512.png': { size: 51484 },
			'apple-touch-icon.png': { size: 12746 },
			'favicon-16x16.png': { size: 626 },
			'favicon-32x32.png': { size: 1487 },
			'mstile-150x150.png': { size: 9050 }
		}
	},
	'polyfills.chunk.*.js': { size: 4066 },
	'push-manifest.json': { size: 303 },
	'favicon.ico': { size: 15086 },
	'manifest.json': { size: 426 },
	'sw.js': { size: 3905 }
};

const expectedOutputs = prepare({
	empty: {
		...smallBuild,
		'bundle.js': { size: 10694 },
		'index.html': { size: 534 },
		'style.css': { size: 131 },
		'style.css.map': { size: 359 },
	},
	simple: {
		...smallBuild,
		'bundle.js': { size: 11336 },
		'index.html': { size: 548 },
		'style.css': { size: 296},
		'style.css.map': { size: 621 },
	},
	root: {
		...fullBuild,
		'bundle.js': { size: 18739 },
		'route-home.chunk.*.js': { size: 959 },
		'route-profile.chunk.*.js': { size: 1595 },
		'index.html': { size: 775 },
		'style.css': { size: 1065 },
		'style.css.map': { size: 2246 },
	},
	'default': {
		...fullBuild,
		'bundle.js': { size: 19661 },
		'route-home.chunk.*.js': { size: 961 },
		'route-profile.chunk.*.js': { size: 1597 },
		'index.html': { size: 775 },
		'style.css': { size: 1065 },
		'style.css.map': { size: 2345 },
	}
});
