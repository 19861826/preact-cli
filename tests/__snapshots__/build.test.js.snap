// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`preact build should build generated projects without errors. Veryfing default 1`] = `
Object {
  "assets": Object {
    "favicon.ico": Object {
      "isFile": true,
      "size": 15086,
    },
    "icons": Object {
      "android-chrome-192x192.png": Object {
        "isFile": true,
        "size": 14058,
      },
      "android-chrome-512x512.png": Object {
        "isFile": true,
        "size": 51484,
      },
      "apple-touch-icon.png": Object {
        "isFile": true,
        "size": 12746,
      },
      "favicon-16x16.png": Object {
        "isFile": true,
        "size": 626,
      },
      "favicon-32x32.png": Object {
        "isFile": true,
        "size": 1487,
      },
      "isDirectory": true,
      "mstile-150x150.png": Object {
        "isFile": true,
        "size": 9050,
      },
    },
    "isDirectory": true,
  },
  "bundle.js": Object {
    "isFile": true,
    "size": 19659,
  },
  "favicon.ico": Object {
    "isFile": true,
    "size": 15086,
  },
  "index.html": Object {
    "isFile": true,
    "size": 775,
  },
  "manifest.json": Object {
    "isFile": true,
    "size": 426,
  },
  "polyfills.chunk.bafcb.js": Object {
    "isFile": true,
    "size": 4068,
  },
  "push-manifest.json": Object {
    "isFile": true,
    "size": 303,
  },
  "route-home.chunk.25be6.js": Object {
    "isFile": true,
    "size": 959,
  },
  "route-profile.chunk.498ba.js": Object {
    "isFile": true,
    "size": 1597,
  },
  "style.css": Object {
    "isFile": true,
    "size": 1065,
  },
  "style.css.map": Object {
    "isFile": true,
    "size": 2345,
  },
  "sw.js": Object {
    "isFile": true,
    "size": 3905,
  },
}
`;

exports[`preact build should build generated projects without errors. Veryfing empty 1`] = `
Object {
  "assets": Object {
    "favicon.ico": Object {
      "isFile": true,
      "size": 15086,
    },
    "icon.png": Object {
      "isFile": true,
      "size": 51484,
    },
    "isDirectory": true,
  },
  "bundle.js": Object {
    "isFile": true,
    "size": 10694,
  },
  "favicon.ico": Object {
    "isFile": true,
    "size": 15086,
  },
  "index.html": Object {
    "isFile": true,
    "size": 534,
  },
  "manifest.json": Object {
    "isFile": true,
    "size": 298,
  },
  "polyfills.chunk.bbde4.js": Object {
    "isFile": true,
    "size": 4068,
  },
  "push-manifest.json": Object {
    "isFile": true,
    "size": 2,
  },
  "style.css": Object {
    "isFile": true,
    "size": 131,
  },
  "style.css.map": Object {
    "isFile": true,
    "size": 359,
  },
  "sw.js": Object {
    "isFile": true,
    "size": 3378,
  },
}
`;

exports[`preact build should build generated projects without errors. Veryfing root 1`] = `
Object {
  "assets": Object {
    "favicon.ico": Object {
      "isFile": true,
      "size": 15086,
    },
    "icons": Object {
      "android-chrome-192x192.png": Object {
        "isFile": true,
        "size": 14058,
      },
      "android-chrome-512x512.png": Object {
        "isFile": true,
        "size": 51484,
      },
      "apple-touch-icon.png": Object {
        "isFile": true,
        "size": 12746,
      },
      "favicon-16x16.png": Object {
        "isFile": true,
        "size": 626,
      },
      "favicon-32x32.png": Object {
        "isFile": true,
        "size": 1487,
      },
      "isDirectory": true,
      "mstile-150x150.png": Object {
        "isFile": true,
        "size": 9050,
      },
    },
    "isDirectory": true,
  },
  "bundle.js": Object {
    "isFile": true,
    "size": 18741,
  },
  "favicon.ico": Object {
    "isFile": true,
    "size": 15086,
  },
  "index.html": Object {
    "isFile": true,
    "size": 775,
  },
  "manifest.json": Object {
    "isFile": true,
    "size": 426,
  },
  "polyfills.chunk.bbde4.js": Object {
    "isFile": true,
    "size": 4068,
  },
  "push-manifest.json": Object {
    "isFile": true,
    "size": 303,
  },
  "route-home.chunk.347f4.js": Object {
    "isFile": true,
    "size": 959,
  },
  "route-profile.chunk.bb3a3.js": Object {
    "isFile": true,
    "size": 1595,
  },
  "style.css": Object {
    "isFile": true,
    "size": 1065,
  },
  "style.css.map": Object {
    "isFile": true,
    "size": 2246,
  },
  "sw.js": Object {
    "isFile": true,
    "size": 3905,
  },
}
`;

exports[`preact build should build generated projects without errors. Veryfing simple 1`] = `
Object {
  "assets": Object {
    "favicon.ico": Object {
      "isFile": true,
      "size": 15086,
    },
    "icon.png": Object {
      "isFile": true,
      "size": 51484,
    },
    "isDirectory": true,
  },
  "bundle.js": Object {
    "isFile": true,
    "size": 11336,
  },
  "favicon.ico": Object {
    "isFile": true,
    "size": 15086,
  },
  "index.html": Object {
    "isFile": true,
    "size": 548,
  },
  "manifest.json": Object {
    "isFile": true,
    "size": 290,
  },
  "polyfills.chunk.bbde4.js": Object {
    "isFile": true,
    "size": 4068,
  },
  "push-manifest.json": Object {
    "isFile": true,
    "size": 2,
  },
  "style.css": Object {
    "isFile": true,
    "size": 296,
  },
  "style.css.map": Object {
    "isFile": true,
    "size": 621,
  },
  "sw.js": Object {
    "isFile": true,
    "size": 3378,
  },
}
`;
